lucet-runtime/lucet-runtime-internals/src/instance/execution.rs

```diff
@@ -125,33 +155,131 @@ pub struct KillState {
+            // not enter the guest region. We will instead terminate the Instance, and then swap
+            // back to the host context.
+            //
+            // Note that we manually drop the domain because`Instance::terminate` never returns.
+            mem::drop(current_domain);
+            (*instance).terminate(TerminationDetails::Remote);
+        }
+    }
+}
+
+/// Exit a guest region.
+///
+/// This is the backstop callback function installed on the [`Instance`](struct.Instance.html) for a
+/// guest. This is called by `lucet_context_backstop` after we have finishing execution in
+/// a guest region.
+///
+/// # Safety
+///
+/// For more information about the safety constraints of the backstop callback, see
+/// [`Instance::init`](struct.Instance.html#method.init).
+///
+/// This function will panic if the `Instance`'s execution domain is marked as pending, currently
+/// in a hostcall, or as cancelled.  Any of these domains mean that something has gone seriously
+/// wrong, and leaving the execution domain mutex in a poisoned state is the least of our concerns.
 pub unsafe extern "C" fn exit_guest_region(instance: *mut Instance) {
     let terminable = (*instance)
         .kill_state
         .terminable
         .swap(false, Ordering::SeqCst);
     if !terminable {
-        // Something else has taken the terminable flag, so it's not safe to actually exit a
-        // guest context yet. Because this is called when exiting a guest context, the
-        // termination mechanism will be a signal, delivered at some point (hopefully soon!).
-        // Further, because the termination mechanism will be a signal, we are constrained to
-        // only signal-safe behavior.
-        //
-        // For now, hang indefinitely, waiting for the sigalrm to arrive.
+        // If we are here, something else has taken the terminable flag, so it's not safe to
+        // actually exit a guest context yet. Because this is called when exiting a guest context,
+        // the termination mechanism will be a signal, delivered at some point (hopefully soon!).
+        // Further, because the termination mechanism will be a signal, we are constrained to only
+        // signal-safe behavior. So, we will hang indefinitely waiting for the sigalrm to arrive.
         #[allow(clippy::empty_loop)]
         loop {}
+    } else {
+        let current_domain = (*instance).kill_state.execution_domain.lock().unwrap();
+        match *current_domain {
+            Domain::Guest => {
+                // We finished executing the code in our guest region normally! We should reset
+                // the kill state, invalidating any existing killswitches' weak references. We
+                // forget the mutex guard so that we don't invoke its destructor and segfault.
+                (*instance).kill_state = Arc::new(KillState::default());
+                mem::forget(current_domain);
```

Comment:
I think you mean `mem::drop` here? I'm pretty sure `forget`ing the lock guard would result in old `KillState` accumulating over time.